library/fsio.c

```diff
@@ -0,0 +1,395 @@
+    CHECK_STATUS( mbedtls_serialize_push_int32( size ), 0, NULL );
+    CHECK_STATUS( mbedtls_serialize_execute( MBEDTLS_SERIALIZE_FUNCTION_FGETS ), 0, NULL );
+    CHECK_STATUS( mbedtls_serialize_pop_buffer( s, size, &len ), 0, NULL );
+    return( s );
+}
+
+/**
+ * \brief          Sets file position. Follows standard C fseek interface.
+ *
+ * \param stream   File handle of type mbedtls_file_t.
+ * \param offset   Offset from whence.
+ * \param whence   Position from where offset is applied.
+ *                 Value is one of MBEDTLS_SEEK_SET, MBEDTLS_SEEK_CUR, or MBEDTLS_SEEK_END.
+ *
+ * \return         returns 0 on success, and -1 on error
+ */
+int mbedtls_fseek( mbedtls_file_t stream, long offset, int whence )
+{
+    switch( whence )
+    {
+        case MBEDTLS_SEEK_SET:
+            whence = MBEDTLS_SERIALIZE_FSEEK_SET;
+            break;
+        case MBEDTLS_SEEK_CUR:
+            whence = MBEDTLS_SERIALIZE_FSEEK_CUR;
+            break;
+        case MBEDTLS_SEEK_END:
+            whence = MBEDTLS_SERIALIZE_FSEEK_END;
+            break;
+        default:
+            return( -1 );
+    }
+    CHECK_STATUS( mbedtls_serialize_push_int32( stream ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_push_int32( whence ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_push_int32( offset ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_execute( MBEDTLS_SERIALIZE_FUNCTION_FSEEK ), 0, -1 );
+    return( 0 );
+}
+
+/**
+ * \brief          Gives current position of file in bytes.
+ *                 Follows standard C ftell interface.
+ *
+ * \param stream   File handle of type mbedtls_file_t.
+ *
+ * \return         returns current position on success, and -1 on error
+ */
+long mbedtls_ftell( mbedtls_file_t stream )
+{
+    long pos = 0;
+
+    CHECK_STATUS( mbedtls_serialize_push_int32( stream ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_execute( MBEDTLS_SERIALIZE_FUNCTION_FTELL ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_pop_int32( (uint32_t *)&pos ), 0, -1 );
+    return( pos );
+}
+
+/**
+ * \brief          Close file. Follows standard C fread interface.
+ *
+ * \param stream   File handle of type mbedtls_file_t.
+ *
+ * \return         Pointer to mbedtls_file_t on success or NULL on failure.
+ */
+int mbedtls_fclose( mbedtls_file_t stream )
+{
+    CHECK_STATUS( mbedtls_serialize_push_int32( stream ), 0, -1 );
+    CHECK_STATUS( mbedtls_serialize_execute( MBEDTLS_SERIALIZE_FUNCTION_FCLOSE ), 0, -1 );
+    return( 0 );
+}
+
+/**
+ * \brief          Test error indicator. Follows standard C ferror interface.
+ *
+ * \param stream   File handle of type mbedtls_file_t.
+ *
+ * \return         Non zero error code if error is set. 0 for no error.
+ */
+int mbedtls_ferror( mbedtls_file_t stream )
+{
+    int status;
+
+    CHECK_STATUS( mbedtls_serialize_push_int32( stream ), 0, -1 );
+    CHECK_STATUS( status = mbedtls_serialize_execute( MBEDTLS_SERIALIZE_FUNCTION_FERROR ), 0, status );
+    return( 0 );
```

Comment:
Why are we potentially returning the `status` error code here but not above?

Reply:
Please check definitions of ```CHECK_STATUS```. Function ```MBEDTLS_SERIALIZE_FUNCTION_FERROR ``` is for fetching file error code. Instead of using a separate parameter, function return code itself is used for getting file error from the frontend. Hence is file error is set then ```mbedtls_serialize_execute``` will return ```!= 0``` status. Check status will return status in that case. Otherwise ```0``` is returned.